Src/Newtonsoft.Json/JsonReader.cs

```diff
@@ -548,81 +519,66 @@ internal byte[] ReadAsBytesInternal()
-                {
-                    byte[] data = ((Guid)Value).ToByteArray();
-                    SetToken(JsonToken.Bytes, data, false);
-                    return data;
-                }
+                case JsonToken.String:
+                    {
+                        // attempt to convert possible base 64 or GUID string to bytes
+                        // GUID has to have format 00000000-0000-0000-0000-000000000000
+                        string s = (string)Value;
 
-                return (byte[])Value;
-            }
+                        byte[] data;
 
-            if (t == JsonToken.StartArray)
-            {
-                List<byte> data = new List<byte>();
+                        Guid g;
+                        if (s.Length == 0)
+                        {
+                            data = new byte[0];
+                        }
+                        else if (ConvertUtils.TryConvertGuid(s, out g))
+                        {
+                            data = g.ToByteArray();
+                        }
+                        else
+                        {
+                            data = Convert.FromBase64String(s);
+                        }
 
-                while (ReadInternal())
-                {
-                    t = TokenType;
-                    switch (t)
+                        SetToken(JsonToken.Bytes, data, false);
+                        return data;
+                    }
+                case JsonToken.Null:
+                case JsonToken.EndArray:
+                    return null;
+                case JsonToken.Bytes:
+                    if (ValueType == typeof(Guid))
                     {
-                        case JsonToken.Integer:
-                            data.Add(Convert.ToByte(Value, CultureInfo.InvariantCulture));
-                            break;
-                        case JsonToken.EndArray:
-                            byte[] d = data.ToArray();
-                            SetToken(JsonToken.Bytes, d, false);
-                            return d;
-                        case JsonToken.Comment:
-                            // skip
-                            break;
-                        default:
-                            throw JsonReaderException.Create(this, "Unexpected token when reading bytes: {0}.".FormatWith(CultureInfo.InvariantCulture, t));
+                        byte[] data = ((Guid)Value).ToByteArray();
+                        SetToken(JsonToken.Bytes, data, false);
+                        return data;
                     }
-                }
 
-                throw JsonReaderException.Create(this, "Unexpected end when reading bytes.");
-            }
+                    return (byte[])Value;
+                case JsonToken.StartArray:
+                    List<byte> buffer = new List<byte>();
 
-            if (t == JsonToken.EndArray)
-            {
-                return null;
+                    for(;;)
```

Comment:
Just that I've always used an empty `for`. I'll change it to match your style. I've a few other comments to make, but something came up just after that last push.